src/ports/postgres/modules/deep_learning/madlib_keras_automl.py_in

```diff
@@ -500,63 +648,432 @@ class KerasAutoML():
+            except ValueError:
+                pass
+        _assert(len(automl_params_dict) >= 1 and len(automl_params_dict) <= 3,
+                "DL: Only num_models, num_iters, and algorithm may be specified")
+        for i in automl_params_dict:
+            if i == AutoMLSchema.NUM_MODELS:
+                self.num_models = automl_params_dict[AutoMLSchema.NUM_MODELS]
+            elif i == AutoMLSchema.NUM_ITERS:
+                self.num_iters = automl_params_dict[AutoMLSchema.NUM_ITERS]
+            elif i == AutoMLSchema.ALGORITHM:
+                if automl_params_dict[AutoMLSchema.ALGORITHM].lower() == 'rand':
+                    self.algorithm = rand
+                elif automl_params_dict[AutoMLSchema.ALGORITHM].lower() == 'tpe':
+                    self.algorithm = tpe
+                # elif automl_params_dict[AutoMLSchema.ALGORITHM].lower() == 'atpe':
+                #     self.algorithm = atpe
+                # uncomment the above lines after atpe works
+                else:
+                    plpy.error("DL: algorithm in 'automl_params' must be 'rand', 'tpe'") # , or 'atpe'
+            else:
+                plpy.error("DL: {0} is an invalid automl param".format(i))
+        _assert(self.num_models > 0 and self.num_iters > 0, "DL: num_models and num_iters in 'automl_params' "
+                                                            "must be positive")
+        _assert(self._is_valid_metrics_compute_frequency(self.num_iters), "DL: 'metrics_compute_frequency' "
+                                                                          "out of iteration range")
+
+    def find_hyperopt_config(self):
         """
-        Drops and re-create the mst table to only include the best performing model configuration.
+        Executes hyperopt on top of MOP.
         """
-        drop_tables([self.model_selection_table])
 
-        # only retaining best performing config
-        plpy.execute("CREATE TABLE {self.model_selection_table} AS SELECT mst_key, model_id, compile_params, " \
-                     "fit_params FROM {self.model_info_table} " \
-                     "ORDER BY {AutoMLSchema.LOSS_METRIC} LIMIT 1".format(self=self, AutoMLSchema=AutoMLSchema))
+        make_mst_summary = True
+        trials = Trials()
+        domain = Domain(None, self.get_search_space())
+        rand_state = np.random.RandomState(self.random_state)
+        configs_lst = self.get_configs_list()
 
-    def generate_model_output_summary_table(self, model_training):
+        self.start_training_time = self.get_current_timestamp()
+        for low, high in configs_lst:
+            i, n = low, high - low + 1
+
+            # Using HyperOpt TPE/ATPE to generate parameters
+            hyperopt_params = []
+            sampled_params = []
+            for j in range(i, i + n):
+                new_param = self.algorithm.suggest([j], domain, trials, rand_state.randint(0, 2 ** 31 - 1))
```

Comment:
I don't think we should calculate 2**31 every time we run this. Setting an INT_MAX constant would be better.